File: src/sources/twitter.js
const SEARCH_TERMS = [
  "AI agent",
  "tool calling",
  "coding agent",
  "MCP server",
  "function calling",
];

export async function fetchTwitter() {
  const apiKey = process.env.SOCIALDATA_API_KEY;
  if (!apiKey) {
    console.warn("SOCIALDATA_API_KEY not set, skipping Twitter");
    return [];
  }

  const results = [];

  for (const term of SEARCH_TERMS) {
    try {
      const query = `${term} min_faves:50`;
      const params = new URLSearchParams({
        query,
        type: "Latest",
      });
      const res = await fetch(
        `https://api.socialdata.tools/twitter/search?${params}`,
        { headers: { Authorization: `Bearer ${apiKey}` } }
      );
      if (!res.ok) continue;
      const data = await res.json();

      for (const tweet of data.tweets || []) {
        const likes = tweet.favorite_count || 0;
        const retweets = tweet.retweet_count || 0;
        const engagement = likes + retweets;

        const createdAt = new Date(tweet.created_at);
        if (Date.now() - createdAt.getTime() > 24 * 60 * 60 * 1000) continue;

        const username = tweet.user?.screen_name || "unknown";
        const url = `https://x.com/${username}/status/${tweet.id_str}`;
        const text = tweet.full_text || tweet.text || "";
        const title =
          text.length > 140 ? text.slice(0, 140) + "..." : text;

        results.push({ title, url, source: "Twitter", engagement });
      }
    } catch {
      // skip failed searches
    }
  }

  const seen = new Set();
  return results
    .filter((r) => {
      if (seen.has(r.url)) return false;
      seen.add(r.url);
      return true;
    })
    .sort((a, b) => b.engagement - a.engagement)
    .slice(0, 20);
}

File: ai_daily_digest/sources/twitter.py
import os
import httpx
from datetime import datetime, timedelta, timezone

SEARCH_TERMS = [
    "AI agent",
    "tool calling",
    "coding agent",
    "MCP server",
    "function calling",
]


def fetch_twitter():
    api_key = os.environ.get("SOCIALDATA_API_KEY")
    if not api_key:
        print("SOCIALDATA_API_KEY not set, skipping Twitter")
        return []

    since = datetime.now(timezone.utc) - timedelta(hours=24)
    results = []

    for term in SEARCH_TERMS:
        try:
            query = f"{term} min_faves:50"
            params = {"query": query, "type": "Latest"}
            res = httpx.get(
                "https://api.socialdata.tools/twitter/search",
                params=params,
                headers={"Authorization": f"Bearer {api_key}"},
                timeout=15,
            )
            if res.status_code != 200:
                continue
            data = res.json()

            for tweet in data.get("tweets", []):
                likes = tweet.get("favorite_count", 0)
                retweets = tweet.get("retweet_count", 0)
                engagement = likes + retweets

                created_at = datetime.strptime(
                    tweet["created_at"], "%a %b %d %H:%M:%S %z %Y"
                )
                if created_at < since:
                    continue

                username = tweet.get("user", {}).get("screen_name", "unknown")
                url = f"https://x.com/{username}/status/{tweet.get('id_str', '')}"
                text = tweet.get("full_text") or tweet.get("text", "")
                title = text[:140] + "..." if len(text) > 140 else text

                results.append(
                    {"title": title, "url": url, "source": "Twitter", "engagement": engagement}
                )
        except Exception:
            continue

    seen = set()
    deduped = []
    for r in results:
        if r["url"] not in seen:
            seen.add(r["url"])
            deduped.append(r)
    deduped.sort(key=lambda r: r["engagement"], reverse=True)
    return deduped[:20]
